src/toxcore/dht_new/server.rs

```diff
@@ -0,0 +1,525 @@
+use tokio_io::IoFuture;
+
+use std::io::{ErrorKind, Error};
+use std::net::SocketAddr;
+use std::collections::HashMap;
+
+use toxcore::crypto_core::*;
+use toxcore::dht_new::packet::*;
+use toxcore::dht_new::packed_node::*;
+use toxcore::dht_new::kbucket::*;
+use toxcore::dht_new::client::*;
+use toxcore::dht_new::codec::*;
+
+/// Type representing Dht UDP packets.
+//pub type DhtUdpPacket = (SocketAddr, DhtPacket);
+/// Shorthand for the transmit half of the message channel.
+type Tx = mpsc::UnboundedSender<DhtUdpPacket>;
+
+/**
+Own DHT node data.
+
+Contains:
+
+- DHT public key
+- DHT secret key
+- Close List ([`Kbucket`] with nodes close to own DHT public key)
+- ping timeout lists ([`TimeoutQueue`])
+
+# Adding node to Close List
+
+Before a [`PackedNode`] is added to the Close List, it needs to be
+checked whether:
+
+- it can be added to [`Kbucket`] \(using [`Kbucket::can_add()`])
+- [`PackedNode`] is actually online
+
+Once the first check passes node is added to the temporary list, and
+a [`NodesRequest`] request is sent to it in order to check whether it's
+online. If the node responds correctly within [`PING_TIMEOUT`], it's
+removed from temporary list and added to the Close List.
+
+[`NodesRequest`]: ../dht/struct.NodesRequest.html
+[`Kbucket`]: ../dht/struct.Kbucket.html
+[`Kbucket::can_add()`]: ../dht/struct.Kbucket.html#method.can_add
+[`PackedNode`]: ../dht/struct.PackedNode.html
+[`PING_TIMEOUT`]: ../timeout/constant.PING_TIMEOUT.html
+[`TimeoutQueue`]: ../timeout/struct.TimeoutQueue.html
+*/
+#[derive(Clone)]
+pub struct Server {
+    /// secret key
+    pub sk: SecretKey,
+    /// public key
+    pub pk: PublicKey,
+    /// Close List (contains nodes close to own DHT PK)
+    pub kbucket: Kbucket,
+}
+
+impl Server {
+    /**
+    Create new `Server` instance.
+
+    Note: a new instance generates new DHT public and secret keys.
+
+    DHT `PublicKey` and `SecretKey` are supposed to be ephemeral.
+    */
+    pub fn new() -> Result<Server, Error> {
+        if !crypto_init() {
+            return Err(Error::new(ErrorKind::Other,
+                       "Crypto initialization failed."));
+        }
+
+        let (pk, sk) = gen_keypair();
+        let kbucket = Kbucket::new(KBUCKET_BUCKETS, &pk);
+
+        debug!("Created new Server instance");
+        Ok(Server {
+            sk: sk,
+            pk: pk,
+            kbucket: kbucket,
+        })
+    }
+
+    /// get pk of DhtPacket
+    pub fn get_pk(packet: &DhtPacket) -> PublicKey {
+        match packet.clone() {
+            DhtPacket::PingRequest(ping_request) => ping_request.pk,
+            DhtPacket::PingResponse(ping_response) => ping_response.pk,
+            DhtPacket::NodesRequest(nodes_request) => nodes_request.pk,
+            DhtPacket::NodesResponse(nodes_response) => nodes_response.pk,
+            DhtPacket::DhtRequest(dht_request) => dht_request.spk,
+            _ => unreachable!("Received packet is not a DhtPacket"),
```

Comment:
Actually it's not unreachable because it depends on how this method is called, so panic is preferable in such cases.

Reply:
But why we can't just get pk when we need it from the packet directly?

Reply:
> But why we can't just get pk when we need it from the packet directly?

How?

Reply:
It means that we should get it only after matching on packet kind but not before. So we can try to move the client inside the server and get exact client by pk from packet in each handle_* method if we need it. Not sure how exact it will look like though, but perhaps it worth to try.

Reply:
Good Idea.
But I would like to try this idea on next PR.:)